File: Tries/208. Implement Trie (Prefix Tree)/solution.cpp
class Trie {
private:
    class TrieNode {
    public:
        TrieNode* next[26];
        bool isEnd;

        TrieNode() {
            for (int i=0; i<26; i++) {
                next[i] = NULL;
            }
            isEnd = false;
        }
    };
    TrieNode* root;
public:
    Trie() {
        root = new TrieNode();
    }
    
    void insert(string word) {
        TrieNode* node = root;
        for (auto& c : word) {
            int key = c-'a';
            if (node->next[key] == NULL) {
                node->next[key] = new TrieNode();
            }
            node = node->next[key];
        }
        node->isEnd = true;
    }
    
    bool search(string word) {
        TrieNode* node = root;
        for (auto& c : word) {
            if (node->next[c-'a'] == NULL) return false;
            node = node->next[c-'a'];
        }
        return node->isEnd;
    }
    
    bool startsWith(string prefix) {
        TrieNode* node = root;
        for (auto& c : prefix) {
            if (node->next[c-'a'] == NULL) return false;
            node = node->next[c-'a'];
        }
        return true;
    }
};

File: Tries/208. Implement Trie (Prefix Tree)/solution.py
class TrieNode:
    def __init__(self):
        self.next = {}
        self.eow = False # end of word

class Trie:

    def __init__(self):
        self.root = TrieNode()
        

    def insert(self, word: str) -> None:
        node = self.root
        for ch in word:
            if ch not in node.next:
                node.next[ch] = TrieNode()
            node = node.next[ch]
        node.eow = True
        

    def search(self, word: str) -> bool:
        node = self.root
        for ch in word:
            if ch not in node.next:
                return False
            node = node.next[ch]
        return node.eow
        

    def startsWith(self, prefix: str) -> bool:
        node = self.root
        for ch in prefix:
            if ch not in node.next:
                return False
            node = node.next[ch]
        return True


# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)